File: warm_ups/wk_07/Day 4 - Simple sums (both)/js/main.js
// ## Simple Sums 

// Write a program that can calculate the sum of the first `n` elements of the following sequences:

// ```js
// s1 = 1 - 1 + 1 - 1 + 1 - 1 + //etc
// s2 = 1 + 2 + 3 + 4 + 5 + 6 + //etc
// ```
// For example:

// ```js
// s1(4) = 0
// s2(4) = 10
// ```

// Use any language of your choice.

// Can you guess what the sum would be if n is infinity?

// ---------------------------------------------------------------

// var s1 = function( num ){
//   output = 1;
//   for( var i = 0; i < num; i++ ){
//     if( i % 2 === 0 ){
//       output --;
//     } else {
//       output ++;
//     }
//   }
//   return output;
// };

// You may notice a pattern, we don't need to loop, we can just check if the number is even or not.

// var s1 = function( num ){
//   var output = 1;
//     if( num % 2 === 0 ){
//       output = 0;
//     } else {
//       output = 1;
//     }
//   return output;
// };

// We can go further:

// Ternary operators behave like extremely short if/else statements. 
// They follow consistent structure like this:
// ( condition ) ? (true case) : (false case)

// If the condition is true, it runs the first block, if not, it does the second.


var s1 = function( num ){
  return num % 2 === 0 ? 0 : 1;
};

console.log(s1(10));
console.log(s1(9));

// ---------------------------------------------------------------


// var s2 = function( num ){
//   var previousStep = 0;
//   for( var i = 1; i <= num; i++ ){
//     previousStep += i;
//   }
//   return previousStep;
// };


// Though the one above works, you may have noticed there's also a pattern here.
// This pattern is so common, it has its own wikipedia page: https://en.wikipedia.org/wiki/1_%2B_2_%2B_3_%2B_4_%2B_%E2%8B%AF

// From which we can steal the following formula:

var s2 = function( num ){
  return ( num * ( num + 1 ) ) / 2;
};

console.log(s2(4));

File: warm_ups/wk_07/Day 4 - Simple sums (both)/main.rb
# Simple Sums

# Write a program that can calculate the sum of the first n elements of the following sequences:

# s1 = 1 - 1 + 1 - 1 + 1 - 1 + //etc
# s2 = 1 + 2 + 3 + 4 + 5 + 6 + //etc
# For example:

# s1(4) = 0
# s2(4) = 10
# Use any language of your choice.

# Can you guess what the sum would be if n is infinity?

class SimpleSums

  def initialize num
    @num = num
  end

  # 1 - 1 + 1 - 1 + 1 - 1 + //etc

  # Using an 'upto' loop because I want to start my loop from 1, not 0. 'times' would start from 0.

  # def first_sum
  #   output = 1
  #   1.upto(@num) do | i |
  #     if i % 2 == 0
  #       output -= 1
  #     else
  #       output += 1
  #     end
  #   end
  #   output
  # end

  # You may have noticed a pattern here with the loop, you can just check whether the number is odd or even

  # def first_sum
  #   output = 1
  #     if @num % 2 == 0
  #       output = 0
  #     else
  #       output = 1
  #     end
  #   output
  # end

  # Ternary operators behave like extremely short if/else statements. 
  # They follow consistent structure like this:
  # ( condition ) ? (true case) : (false case)

  # If the condition is true, it runs the first block, if not, it does the second.

  def first_sum
    @num % 2 == 0 ? 0 : 1
  end

  # 1 + 2 + 3 + 4 + 5 + 6

  # I'm using output to store the previous step in the cycle and += to add it to i.

  # def second_sum
  #   output = 0;
  #   1.upto( @num ) do | i |
  #     output = output + i
  #   end
  #   output
  # end

  def second_sum
    ( @num * (@num + 1) ) / 2
  end

end


sum = SimpleSums.new( 4 )

p sum.first_sum
p sum.second_sum


# sum = SimpleSums.new( Float::INFINITY )
# => 1